flo-workflow/src/main/java/com/spotify/flo/TaskOperator.java

```diff
@@ -20,22 +20,15 @@
 
 package com.spotify.flo;
 
-import com.spotify.flo.context.OperatingContext;
 import java.io.Serializable;
 
 /**
  * An operator controls the execution of a job for a task,  e.g. a data processing job on some processing platform.
  *
  * <p>The concrete operator implementation should {@link #provide(EvalContext)} the task with some means of constructing
- * an operation description. The operation description should be thrown out of the process fn using an
- * {@link OperationException} to be caught and executed by the {@link OperatingContext}.
+ * an operation description. The operation description should be returned from the process fn.
  */
-public abstract class TaskOperator<T> extends TaskContextGeneric<T> {
-
-  public abstract static class OperationException extends ControlException {
-
-    public abstract <T> T run(Listener listener);
-  }
+public abstract class TaskOperator<T, Y, Z> extends TaskContextGeneric<T> {
```

Comment:
I noticed a potential pitfall with these type arguments when implementing a `TaskOperator` which lead to user code written using that operator cause a `ClassCastException` even though the code compiled. The changes in `TaskBuilder` do a good job at keeping `T` and `Y` in check since they are used as the input arg type and return type of `process`. However, there is no tie between `Y` and `Z` which means that the type used for `Y` (the jobspec) might be implemented in a way so that it would compile something that return something other than `Z` to `Z perform(Y o, Listener listener)` in `TaskOperator`.

I might not be explaining this very clearly so here's an example using `JobOperator` from the test.

```java
Task<JobResult> task = Task.named("task", today).ofType(JobResult.class)
    .context(JobOperator.create())
    .process(job -> job
        .options(() -> ImmutableMap.of("quux", 17))
        .pipeline(ctx -> ctx.readFrom("foo").map("x + y").writeTo("baz"))
        .validation(result -> {
          if (result.records < 5) {
            throw new AssertionError("Too few records seen!");
          }
        })
        .success(result -> "oh a string!" /* should be JobResult */));
```

This compiles fine, but causes a `ClassCastException`.

I think an interface that would allow `Y` and `Z` to be related could remedy this. But it might restrict how the job spec can be implemented. I'm thinking something like:

```java
public abstract class TaskOperator<T, Y extends OperatorSpec<Z>, Z> extends TaskContextGeneric<T> {

  public interface OperatorSpec<ZZ> {
    ZZ success(/* need freedom here */);
  }

  // ...
}
```

I realise this is mainly a pitfall that operator implementors might run into that then can cause confusion for users. So maybe it can just be documented as something to be aware of in the `TaskOperator` javadoc.

Reply:
Suggested change: https://github.com/spotify/flo/commit/96dc47042973dba923d871401dd8fdda938a6dbf